### PR TITLE
several changes using introspection advection field,  degree and component index

### PR DESCRIPTION
**Pull request by yinghe616**
When go though the advection solver part,  I found several places can be improved by using current introspection properties for advection field.

Besides above, I also made one change on assembly.cc . For DG approach, the current Dirichlet boundary condition seems only be weakly imposed on inflow boundary condition no matter the filed is temperature field or compositional filed, but I think we don't have to check this for the temperature filed. So I'll just wait for other opinions.

### CODE REVIEW

File: source/simulator/assembly.cc [hunk 1]
```diff
@@ -1419,7 +1419,8 @@ namespace aspect
                    * on the direction of the flow. Thus the flow-dependent terms below are
                    * only calculated if the edge is an inflow edge.
                    */
-                  const bool inflow = ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.);
+                  const bool inflow = (advection_field.is_temperature()
+                                       || ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.));
```
Comment (bangerth) [inline source/simulator/assembly.cc]: Can you explain this change in more detail? It makes all boundaries inflow boundaries if we are considering the temperature. Maybe that's what we want, but I don't understand why this would be so.

**Comment by bangerth** [comment]
Nice job finding these places -- making code easier is always great. But I don't understand the first of the three changes with the inflow boundary. I look forward to an explanation :-)

File: source/simulator/assembly.cc (hunk 1 above)
Comment (yinghe616) [inline source/simulator/assembly.cc]: I just thought this from PDE. The temperature equation is a second order equation in spatial, so we could impose entire Dirichlet boundary condition, but compositional equation is a first order equation in spatial, so we can only impose the inflow Dirichlet boundary condition. However, the current code seems only impose inflow boundary condition, then for temperature which may have also imposed outflow boundary condition (which I am not sure if physically practical), the code will impose nothing on the outflow boundary, which then becomes a weakly imposed zero Neumman boundary condition.

Comment (bangerth) [inline source/simulator/assembly.cc]: Hm, I think this requires a bit more study. I definitely see that there is a question that needs to be answered. Are you willing to remove this part of the patch and put it into a separate pull request? The other parts are uncontroversial and should be merged; they are also logically independent of this issue, and we try to keep logically separate issues in separate pull requests.

Comment (spco) [inline source/simulator/assembly.cc]: I think @yinghe616 is correct here, so long as we don't allow a degenerate diffusion coefficient. This would then also be consistent with the comment in line 1343 above. The hyperbolic nature of the composition equation means it's obviously only sensible to impose Dirichlet on inflow. The non-degenerate (I presume ASPECT never allows zero temperature diffusion?) elliptic temperature equation should just have one of Dirichlet and Neumann imposed at each boundary point, independent of flow direction (and imposing nothing is identical to zero Neumann). 

This code was originally built to handle the case of a degenerate diffusion, my copy-paste-mistake!

Perhaps it would also be better to rename this something other than `inflow`, though I can't currently think of anything much snappier than `on_composition_inflow_boundary_or_advection_field_is_temperature` ;)

Thanks for catching this @yinghe616!

Comment (gassmoeller) [inline source/simulator/assembly.cc]: I would be careful here. Of course in forward calculations the diffusion coefficient is always non-zero, but an often used approach to approximate backward advection is to set the diffusion coefficient to zero (because you can not reverse diffusive processes) and reverse the sign of the gravity. @jaustermann did some tests in that direction during the hackathon, and if it is not much effort then I would prefer keeping the possibility to set up backward advection models with discontinuous elements.

Comment (bangerth) [inline source/simulator/assembly.cc]: Thinking about this some more, shouldn't we treat Dirichlet boundary conditions via the standard explicit jump terms like we do on faces between cells, but where one half of the jump is simply the prescribed boundary term? In other words, specifically *not* use flux terms for Dirichlet values? This would mean doing the *opposite* from what this part of the patch does.

**Comment by yinghe616** [comment]
I remove the DG Dirichlet temperature boundary implementation of the patch and I'll put it into a separate pull request.

File: source/simulator/assembly.cc (hunk 1 above)
Comment (spco) [inline source/simulator/assembly.cc]: Since we are in the convection-dominated regime in general, we should philosophically be imposing the Dirichlet values via the flux terms. The jump terms come more from the pure diffusion elliptic limit, and so are relatively unimportant here. Neglecting the flow jumps would probably be a bad idea.

Comment (bangerth) [inline source/simulator/assembly.cc]: Ah, very relevant observation. Let's continue this conversation in the separate issue.

**Comment by bangerth** [comment]
Ready to merge if the tester is happy.

File: source/simulator/assembly.cc (hunk 1 above)
Comment (yinghe616) [inline source/simulator/assembly.cc]: @bangerth , can you explain more about your question?
"one half of the jump is simply the prescribed boundary term?"

I thought Sam's implementation for Dirichlet boundary seems right which are 

(u\cdot\grad T, \phi)_e  = - (\grad T, \phi u)_e +(u\cdot n T, \phi)_b  =  (u\cdot\grad T, \phi)_e -  (u\cdot n  T ^*, \phi)_b +(u\cdot n T, \phi)_b

on the boundary b, in the second term T^* uses the Dirichlet value and put into the RHS vector contribution, but  in the third term T put into the local matrix contribution
